src/exchange/moderators/truncate.py

```diff
@@ -1,59 +1,70 @@
+from exchange.message import Message
 from exchange.moderators import PassiveModerator
 from exchange.moderators.base import Moderator
 
-MAX_TOKENS = 112000
-
-
-def pop_checkpoint(
-    exchange: Type["exchange.exchange.Exchange"],  # noqa: F821
-    exclude_last: Literal[0, 1] = 0,
-) -> Type["exchange.exchange.Exchange"]:  # noqa: F821
-    """Pop messages from the front of the list in sections"""
-    checkpoint = exchange.checkpoints.pop(0)
-    for _ in range(checkpoint.end_index - checkpoint.start_index - exclude_last):
-        exchange.messages.pop(checkpoint.start_index)
-
-    # rewrite checkpoint indexes
-    for cp in exchange.checkpoints:
-        cp.start_index = cp.start_index - checkpoint.end_index + exclude_last
-        cp.end_index = cp.end_index - checkpoint.end_index + exclude_last
-
-    if exclude_last:
-        # there is still one entry in the first checkpoint only
-        checkpoint.token_count = checkpoint.latest_generated_tokens
-        checkpoint.end_index = exclude_last
-        exchange.checkpoints.insert(0, checkpoint)
-
-    return exchange
+# currently this is the point at which we start to truncate, so
+# so once we get to this token size the token count will exceed this
+# by a little bit
+MAX_TOKENS = 70000
 
 
 class ContextTruncate(Moderator):
-    def __init__(self, model: Optional[str] = None) -> None:
+    def __init__(
+        self,
+        model: Optional[str] = "gpt-4o-mini",
+        max_tokens: Optional[int] = MAX_TOKENS,
+    ) -> None:
         self.model = model
-        self.system_prompt_token_count = None
+        self.system_prompt_token_count = 0
+        self.max_tokens = max_tokens
+        self.last_system_prompt = None
 
-    def rewrite(self, exchange: Type["exchange.exchange.Exchange"]) -> None:  # noqa: F821
+    def rewrite(self, exchange: Type["exchange.exchange.Exchange"]) -> None:
         """Truncate the exchange messages with a FIFO strategy."""
-        if not self.system_prompt_token_count:
+        self._update_system_prompt_token_count(exchange)
+        messages_to_remove = self._get_messages_to_remove(exchange)
+        for _ in range(len(messages_to_remove)):
+            exchange.pop_first_message()
+
+    def _update_system_prompt_token_count(self, exchange: Type["exchange.exchange.Exchange"]) -> None:
+        is_different_system_prompt = False
+        if self.last_system_prompt != exchange.system:
+            is_different_system_prompt = True
+            self.last_system_prompt = exchange.system
+
+        if not self.system_prompt_token_count or is_different_system_prompt:
             # calculate the system prompt tokens (includes functions etc...)
             _system_token_exchange = exchange.replace(
                 messages=[],
-                checkpoints=[],
+                checkpoint_data=CheckpointData(),
                 moderator=PassiveModerator(),
                 model=self.model if self.model else exchange.model,
```

Comment:
This should prob be the same model as the initial exchange so that token content is consistent with what we're expecting. Between models on openai, this is ok, but if we use claude where while the original is gpt-4, there will be discrepancies (don't expect to be major though)

Reply:
Good call, I just added a check for this on the `.replace()` method - any calls to `.replace()` that specify `model` in the `kwargs` should also specify `checkpoint_data`.

I'm not convinced that we should override to usingt he exchange's model here, as it defeats the purpose of specifying an accelerator model in the profile being used. Instead, we expect folks will configure profiles in a meaningful way, and our scripting should add enough token count padding to account for some user errors.

Reply:
I just meant for this specific case to calculate the system message token count, but I guess this effects everything. +1 on users using self consistent profiles

Reply:
> Good call, I just added a check for this on the .replace() method - any calls to .replace() that specify model in the kwargs should also specify checkpoint_data.

I actually ended up removing this, as it causes too many new errors in Goose. This is indicative of us already falling into patterns of mutating the Exchange in potentially unsafe ways.